torch/csrc/jit/tensorexpr/ir_simplifier.cpp

```diff
@@ -2404,6 +2404,92 @@ Stmt* TermExpander::mutate(const Block* v) {
   return fuseSyncThreads(new_block);
 }
 
+// SimplifierUnderContext
+
+Stmt* SimplifierUnderContext::mutate(const For* v) {
+  const Expr* var = v->var();
+  const Expr* start = v->start();
+  const Expr* stop = v->stop();
+  Stmt* body = v->body();
+  LoopOptions loop_options = v->loop_options();
+  const Expr* var_new_expr = var->accept_mutator(this);
+  const Var* var_new = dynamic_cast<const Var*>(var_new_expr);
+  const Expr* start_new = start->accept_mutator(this);
+  const Expr* stop_new = stop->accept_mutator(this);
+  Stmt* body_new = body;
+
+  // save bounds info before this for-stmt
+  bool has_bounds = false;
+  std::pair<const Expr*, const Expr*> bound_old;
+  const Var* var_key = dynamic_cast<const Var*>(var);
+  auto got = var_bound_info_.find(var_key);
+  if (got != var_bound_info_.end()) {
+    has_bounds = true;
```

Comment:
Isn't every loop variable unique, even though their "name" could match? When is this case possible?

Reply:
I was considering more context-sensitive simplifications in nested `cond` and `for` stmts. For example, 
```
if (i>=0 && i<5) {
  for (i=0; i<3; i++){
    A[i] = ...
  }
  x = (i+20) / 5;
}
```
Inside the `if` stmt, `i` is in the range of [0, 5); and if we can restore this bound info after the `for` stmt, we can use it to simplify the assignment stmt `x = (i+20)/5` to `x = 4`.

This (simplifications inside `if` stmt) is not supported yet but I expect to gradually support such cases.

Reply:
Can you please add this explanation as a comment here? Otherwise, it is difficult to understand why we are doing this.